# StatSTR : a tool for computing tandem repeat statistics from VCF files

[Usage](#usage) | [Statistics options](#stats) | [Example command](#example) | [Output files](#outputs) 

StatSTR takes in a VCF file generated by a supported TR genotyping tool (GangSTR, HipSTR, popSTR, ExpansionHunter, AdVNTR) and outputs per-locus statistics.

<a name="usage"></a>
## Usage 
To run statSTR use the following command: 
```
statSTR 
  --vcf <vcf file> \
  --out test.tab \
  --thresh \
  [additional options]
```

Required Parameters: 
* `--vcf`: input the STR VCF file 
* `--vcftype <string>`: Type of VCF file being processed. Default='auto'. Must be one of: 'gangstr', 'advntr', 'hipstr', 'eh', 'popstr'.
* `--out`: prefix to name output files. Set to stdout to write to standard output.

Filtering Group parameters: 
* `--samples`: A file containing a list of samples to include in computing statistics. If not given, all samples are used.
* `--region`: restrict to specific regions (chrom:start-end). 

For specific statistics available, see below.

StatSTR will output a tab delimited file with per-locus statistics. See a description of the output file below.

<a name="stats"></a>
## Statistics options

The following options can be specified to compute various per-locus statistics:

| StatSTR option | Description |
|----| ----|
| `--thres` | Output the maximum observed allele length. (Output column="thresh") |
| `--afreq` | Output allele frequences. (Output column="afreq"). Comma-separated list of allele:freq | 
| `--acount` | Output allele counts. (Output column="acount"). Comma-separated list of allele:count | 
| `--hwep` | Output Hardy Weinberg p-values oer locus. (Output column="hwep") |
| `--het` | Output heterozygosity of each locus. (Output column="het") |
| `--mean` | Output mean allele length. (Output column="mean") |
| `--mode` | Output mode allele length. (Output column="mode") |
| `--var` | Output variance of allele length. (Output column="var") |
| `--numcalled` | Output number of called samples. (Output column="numcalled") |

Specifying `--use-length` will collapse alleles by length (only relevant for HipSTR, which outputs sequence differences in TR alleles).

<a name="example"></a>
## Example command

If you have installed the STRTools package, you can run an example statSTR command using files in this repository.

```
statSTR \
  --vcf $REPODIR/test/common/sample_vcfs/test_hipstr.vcf \
  --out teststats.tab \
  --afreq
```
where `$REPODIR` points to the root path of this repository.

This will output `teststats.tab` with columns chrom, start, end, afreq.

If you are testing from source, you can run:
```
python statSTR.py \
  --vcf $REPODIR/test/common/sample_vcfs/test_hipstr.vcf \
  --out teststats.tab \
  --afreq
```

<a name="outputs"></a>
## Output file

StatSTR outputs a tab-delimited file with columns: chrom, start, end, plus an additional column for each statistic specified using the options described above.

